pkg/operator/rotate.go

```diff
@@ -0,0 +1,242 @@
+	// has an expiry that guarantees a minimum trust duration.
+	var currentCACertExpiry *time.Time
+	minimumExpiry := time.Now().Add(minimumTrustDuration)
+	if currentCACert.NotAfter.Before(minimumExpiry) {
+		currentCACertExpiry = &minimumExpiry
+	}
+
+	// Generate an intermediate cert bridging trust between the new CA and serving certs
+	// generated by the current CA for inclusion in the new CA bundle. This will ensure
+	// that clients with a post-rotation ca bundle will be able to trust pre-rotation
+	// serving certs.
+	currentCACertSignedByNewCA, err := createIntermediateCACert(currentCACert, newCACert, newCAConfig.Key.(*rsa.PrivateKey), currentCACertExpiry)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create intermediate certificate signed by new ca: %v", err)
+	}
+
+	bundle := []*x509.Certificate{
+		newCACert,
+		currentCACertSignedByNewCA,
+	}
+
+	// Generate an intermediate cert bridging trust between the current CA and serving
+	// certs generated by the new CA. This cert will need to be included with serving
+	// certs generated by the new CA to ensure that clients with the pre-rotation ca
+	// bundle will be able to trust post-rotation serving certs.
+	newCACertSignedByOldCA, err := createIntermediateCACert(newCACert, currentCACert, currentKey, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create intermediate certificate signed by previous ca: %v", err)
+	}
+
+	return &signingCA{
+		config:             newCAConfig,
+		bundle:             bundle,
+		intermediateCACert: newCACertSignedByOldCA,
+		oldCAExpiry:        currentCACertSignedByNewCA.NotAfter,
+	}, nil
+}
+
+// createIntermediateCACert creates a new intermediate CA cert from a template provided by
+// the target CA cert and issued by the signing cert. This ensures that certificates
+// issued by the target CA can be trusted by clients that trust the signing CA.
+func createIntermediateCACert(targetCACert, signingCACert *x509.Certificate, signingKey *rsa.PrivateKey, expiry *time.Time) (*x509.Certificate, error) {
+	// Copy the target cert to allow modification.
+	template, err := x509.ParseCertificate(targetCACert.Raw)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to copy ca certificate: %v", err)
+	}
+	// Enable key identity chaining
+	template.AuthorityKeyId = signingCACert.SubjectKeyId
+
+	// Update the expiry if necessary
+	if expiry != nil {
+		template.NotAfter = *expiry
+	}
+
+	caCert, err := util.CreateCertificate(template, signingCACert, targetCACert.PublicKey, signingKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating intermediate CA certificate: %v", err)
+	}
+
+	return caCert, nil
+}
+
+// certHalfwayExpired indicates whether half of the cert validity period has elapsed.
+func certHalfwayExpired(cert *x509.Certificate) bool {
+	halfValidPeriod := cert.NotAfter.Sub(cert.NotBefore).Nanoseconds() / 2
+	halfExpiration := cert.NotBefore.Add(time.Duration(halfValidPeriod) * time.Nanosecond)
```

Comment:
is that `* time.Nanosecond`  really needed? Doesn't Add take a time.Time?

Reply:
I've removed this function in preference to triggering time-based rotation when the current CA's expiry provides less than the minimum trust duration (~6 months) rather than checking that the CA's is more than half-way through its validity period. This is in keeping with how long the previous CA will be trusted. WDYT?

Reply:
it's the same at the end, isn't it?

Reply:
Effectively the same, but I think it's more clear to using the same constant (`minimumTrustDuration`) for both detection that time-based rotation is required and to ensure that the previous CA remains trusted for at least that same duration.